crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs

```diff
@@ -10,88 +11,211 @@ use rand::SeedableRng;
 pub struct Aes256CbcHmacKey {
     /// Uses a pinned heap data structure, as noted in [Pinned heap data][crate#pinned-heap-data]
     pub(crate) enc_key: Pin<Box<GenericArray<u8, U32>>>,
     /// Uses a pinned heap data structure, as noted in [Pinned heap data][crate#pinned-heap-data]
     pub(crate) mac_key: Pin<Box<GenericArray<u8, U32>>>,
 }
 
+impl ConstantTimeEq for Aes256CbcHmacKey {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.enc_key.ct_eq(&other.enc_key) & self.mac_key.ct_eq(&other.mac_key)
+    }
+}
+
+impl PartialEq for Aes256CbcHmacKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+#[derive(Zeroize, Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct XChaCha20Poly1305Key {
+    pub(crate) key_id: [u8; 24],
+    pub(crate) enc_key: Pin<Box<GenericArray<u8, U32>>>,
+}
+
+impl ConstantTimeEq for XChaCha20Poly1305Key {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.enc_key.ct_eq(&other.enc_key)
+    }
+}
+
+impl PartialEq for XChaCha20Poly1305Key {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
 /// A symmetric encryption key. Used to encrypt and decrypt [`EncString`](crate::EncString)
 #[derive(ZeroizeOnDrop, Clone)]
-#[cfg_attr(test, derive(PartialEq))]
 pub enum SymmetricCryptoKey {
     Aes256CbcKey(Aes256CbcKey),
     Aes256CbcHmacKey(Aes256CbcHmacKey),
+    /// Data encrypted by XChaCha20Poly1305Key keys has type
+    /// [`Cose_Encrypt0_B64`](crate::EncString::Cose_Encrypt0_B64)
+    XChaCha20Poly1305Key(XChaCha20Poly1305Key),
 }
 
 impl SymmetricCryptoKey {
-    const KEY_LEN: usize = 32;
-    const MAC_LEN: usize = 32;
-
-    /// Generate a new random [SymmetricCryptoKey]
-    pub fn generate(mut rng: impl rand::RngCore) -> Self {
+    // enc type 0 old static format
+    const AES256_CBC_KEY_LEN: usize = 32;
+    // enc type 2 old static format
+    const AES256_CBC_HMAC_KEY_LEN: usize = 64;
+
+    /// Generate a new random AES256_CBC [SymmetricCryptoKey]
+    ///
+    /// WARNING: This function should only be used with a proper cryptographic RNG. If you do not
+    /// have a good reason for using this function, use
+    /// [SymmetricCryptoKey::make_aes256_cbc_hmac_key] instead.
+    pub(crate) fn make_aes256_cbc_hmac_key_internal(
+        mut rng: impl rand::RngCore + rand::CryptoRng,
+    ) -> Self {
         let mut enc_key = Box::pin(GenericArray::<u8, U32>::default());
         let mut mac_key = Box::pin(GenericArray::<u8, U32>::default());
 
         rng.fill(enc_key.as_mut_slice());
         rng.fill(mac_key.as_mut_slice());
 
-        SymmetricCryptoKey::Aes256CbcHmacKey(Aes256CbcHmacKey { enc_key, mac_key })
+        Self::Aes256CbcHmacKey(Aes256CbcHmacKey { enc_key, mac_key })
+    }
+
+    /// Generate a new random AES256_CBC_HMAC [SymmetricCryptoKey]
+    pub fn make_aes256_cbc_hmac_key() -> Self {
+        let mut rng = rand::thread_rng();
+        Self::make_aes256_cbc_hmac_key_internal(&mut rng)
```

Comment:
Can also remove the mut here:
```suggestion
        let rng = rand::thread_rng();
        Self::make_aes256_cbc_hmac_key_internal(rng)
```